minvws/nl-kat-coordination
# Add create schedule functionality to scheduler api

### Context

Adds the functionality of creating a schedule for an arbitrary task through the scheduler API. This functionality is needed for the upcoming recurring report tasks. It is necessary to specify at what interval this task needs to be executed for an individual report task. A schedule for a task can be created as follows:

```javascript
POST /schedules
(
    "scheduler_id": "id-of-the-scheduler",
    "schedule": "cron-expression",
    "data": {"place": "your", "task": "here"}
}
```

### Changes

* `models/schedule.py` 
  * ~~computed field for deadline; we want to be able to calculate the deadline for the schedule when we create a schedule from the create endpoint. This allows the API client just to supply the cron expression and not do the deadline calculation itself.~~ Update: a post init to calculate the `deadline_at` when it isn't supplied and a schedule is present did the job better with less code.
* `server/`
  * `schedules.py` added create endpoint for schedules, adds validation of `data` by referencing the scheduler with `scheduler_id`, it also generates the `hash` of the contents of the `data` field so this doesn't have to be done by the client.
  * `serializers/schedule.py` adding `ScheduleCreate` model to correctly serialize/validate the client request. Take note that issues are already created to tackle the server API like: https://github.com/minvws/nl-kat-coordination/issues/3162

### QA notes

We're going to create a `Schedule` through the newly created `/schedules` endpoint:

* `make reset`
* Fire up openkat
*  Create an organisation called `dev`. Don't start any scans though!
* `curl -X POST -H "Content-Type: application/json" -d '{"scheduler_id": "boefje-dev", "schedule": "*/5 * * * *", "data": {"id": "12c3f698-5036-40ce-a1d5-1d59e50116d8", "boefje": { "id": "dns-records", "name": "DnsRecords", "version": null }, "input_ooi": "Hostname|internet|mispo.es", "dispatches": [], "organization": "dev" }}' http://localhost:8004/schedules
`
* `curl HTTP://localhost:8004/schedules` it should now list a schedule

---

### Code Checklist







---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.
